solr/core/src/java/org/apache/solr/cli/StreamTool.java

```diff
@@ -0,0 +1,507 @@
+                "Name of the collection to execute on if workers are 'solr'.  Required for 'solr' worker.")
+            .build();
+
+  private static final Option FIELDS_OPTION = Option.builder("f")
+            .longOpt("fields")
+            .argName("FIELDS")
+            .hasArg()
+            .required(false)
+            .desc(
+                "The fields in the tuples to output. (defaults to fields in the first tuple of result set).")
+            .build();
+
+  private static final Option HEADER_OPTION =
+          Option.builder()
+                  .longOpt("header")
+                  .required(false)
+                  .desc("Whether or not to include a header line. (default=false)")
+                  .build();
+  private static final Option DELIMITER_OPTION = Option.builder()
+           .longOpt("delimiter")
+            .argName("CHARACTER")
+            .hasArg()
+            .required(false)
+            .desc("The output delimiter. (default=tab).")
+            .build();
+  private static final Option ARRAY_DELIMITER_OPTION = Option.builder()
+                .longOpt("array-delimiter")
+            .argName("CHARACTER")
+            .hasArg()
+            .required(false)
+            .desc("The delimiter multi-valued fields. (default=|)")
+            .build();
+
+  @Override
+  public Options getOptions() {
+    Options opts =
+            super.getOptions()
+                    .addOption(EXECUTION_OPTION)
+                    .addOption(COLLECTION_OPTION)
+                    .addOption(FIELDS_OPTION)
+                    .addOption(HEADER_OPTION)
+                    .addOption(DELIMITER_OPTION)
+                    .addOption(ARRAY_DELIMITER_OPTION)
+                    .addOption(CommonCLIOptions.CREDENTIALS_OPTION)
+                    .addOptionGroup(getConnectionOptions());
+
+    return opts;
+  }
+
+
+  @Override
+  @SuppressWarnings({"rawtypes"})
+  public void runImpl(CommandLine cli) throws Exception {
+
+    String expressionArgument = cli.getArgs()[0];
+    String execution = cli.getOptionValue(EXECUTION_OPTION,"solr");
+    String arrayDelimiter = cli.getOptionValue(ARRAY_DELIMITER_OPTION,"|");
+    String delimiter = cli.getOptionValue(DELIMITER_OPTION, "   ");
+    boolean includeHeaders = cli.hasOption(HEADER_OPTION);
+    String[] outputHeaders = getOutputFields(cli);
+
+    LineNumberReader bufferedReader = null;
+    String expr;
+    try {
+      Reader inputStream =
+          expressionArgument.toLowerCase(Locale.ROOT).endsWith("expr")
+              ? new InputStreamReader(
+                  new FileInputStream(expressionArgument), Charset.defaultCharset())
+              : new StringReader(expressionArgument);
+
+      bufferedReader = new LineNumberReader(inputStream);
+      expr = StreamTool.readExpression(bufferedReader, cli.getArgs());
+      echoIfVerbose("Running Expression: " + expr);
+    } finally {
+      if (bufferedReader != null) {
+        bufferedReader.close();
+      }
+    }
```

Comment:
There is a lot happening in the `runImpl`, maybe extracting the two blocks into separate private methods would keep a better overview? (this is block no 1)